helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ShardingKeyTrieNode.java

```diff
@@ -0,0 +1,118 @@
+package org.apache.helix.rest.metadatastore;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * A trie node that is used to represent Zk path sharding keys. Terminal nodes contain ZkRealm
+ * addresses.
+ */
+public class ShardingKeyTrieNode {
+  /**
+   * Children trie nodes of this trie node. Each node is identified by their name.
+   */
+  private final Map<String, ShardingKeyTrieNode> _children;
+  /**
+   * Boolean value dictating if this trie node is a terminal node.
+   */
+  private final boolean _isLeaf;
+  /**
+   * This trie node's name.
+   */
+  private final String _name;
+  /**
+   * ZkRealm address corresponding to this trie node. The value is only valid when the node is a
+   * terminal node.
+   */
+  private final String _zkRealmAddress;
+
+  public ShardingKeyTrieNode(Map<String, ShardingKeyTrieNode> children, boolean isLeaf, String name,
+      String zkRealmAddress) {
+    _children = children;
+    _isLeaf = isLeaf;
+    _name = name;
+    _zkRealmAddress = zkRealmAddress;
+  }
+
+  public Map<String, ShardingKeyTrieNode> getChildren() {
+    return Collections.unmodifiableMap(_children);
+  }
+
+  public boolean isLeaf() {
+    return _isLeaf;
+  }
+
+  public String getName() {
+    return _name;
+  }
+
+  public String getZkRealmAddress() {
```

Comment:
Is it the case that this address won't be valid or even initialized if this is not a leaf node? If so, can we create a leaf class which extends this one and make this method available only there?
If this is not the case, please ignore what I said.

Reply:
My instinct is saying a leaf class extension may be an overkill. This is because we can make an argument that the `children` variable won't be valid or even initialized if it's a leaf node, then it seems that we should have two class: `leafNode` and `nonLeafNode`, to avoid the misuse of `children`. `LeafNode` has `zkRealmAddress`, and `nonLeafNode` has `children`; both of these classes inherit from a parent class.
If we look at that purposed parent class, its only variable is `name`, which seems to be too broad for a class to me.

Reply:
What I proposed are TreeNode and LeafNode. LeafNode is the child of the TreeNode. getZkRealmAddress() is the method of LeafNode, TreeNode only supports the tree-related operations. I think it is easier because the caller does not call isLeaf() to check then call getZkRealmAddress() to fetch the information.

For easier access, every TreeNode can have a method called getLeafNodes() which returns all the leaf nodes under its branch.

Then,
1. When you adding a new realm address, construct a LeafNode then adding to the right TreeNode.
2. When you search for a node, find the TreeNode first, then call getLeafNodes() which will return all the leaves.

Reply:
I think it will simplify the class.